<a href="https://colab.research.google.com/github/valperezg02/Analitica-de-Negocios/blob/main/Reto1_ValeriaPerezG.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Reto 1**
En este documento se desarrolla y analiza un modelo Naive Bayes para detección de riesgo de lavado de activos (SARLAFT) en una entidad financiera utilizando las variables socioeconómicas de un solicitante de crédito. De acuerdo con lo anterior, las variables socioeconómicas son:

**Edad:** es la edad que posee el cliente al momento del análisis de riesgo.

**Ingresos_usd:** Ingresos recibidos por una persona como resultado de una actividad comercial.

**Gastos_usd:** Son los gastos del cliente. Gastos de personas a cargo, hijos y servicios públicos.

**Num_tarjetas_credito:** es la cantidad de tarjetas de crédito que el cliente posee.

**Monto_transado_tarjetas_usd:** es la cantidad de dinero que se ha movido en las tarjetas.

**Porcentaje_crecimiento_patrimonio:** indica el crecimiento o disminución del patrimonio del cliente.


0. Se cargan las librerías de trabajo

In [ ]:
import numpy as np
import pandas as pd

from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

1. Se cargan los datos de trabajo de la base de datos LavadoActivos

In [ ]:
nxl='/content/2. LavadoActivos.xlsx'
XDB= pd.read_excel(nxl, sheet_name=0)

# Se seleccionan las variables de trabajo
XD=XDB[['edad','ingresos_usd','gastos_usd', 'num_tarjetas_credito', 'monto_transado_tarjetas_usd','porcentaje_crecimiento_patrimonio']]
XD.head(10)

yd=XDB[['lavado_activos']]
yd.head()

,lavado_activos
0,1
1,1
2,0
3,0
4,0


2. Se implementa el modelo Naive bayes

In [ ]:
mnb=GaussianNB()

# se ajustan las variables entrada - salida
mnb.fit(XD,yd)

# Se muestra la media de las variables
m=mnb.theta_

# Se muestra la Desviación estándar de las variables
sigma=mnb.var_ ; sigma=np.sqrt(sigma)
print('edad','ingresos_usd','gastos_usd', 'num_tarjetas_credito', 'monto_transado_tarjetas_usd','porcentaje_crecimiento_patrimonio')
print(m)
print('Las desviaciones son:')
print(sigma)


edad ingresos_usd gastos_usd num_tarjetas_credito monto_transado_tarjetas_usd porcentaje_crecimiento_patrimonio
[[4.54696356e+01 6.69073255e+03 3.60174970e+03 1.95456590e+00
  3.09046404e+03 4.46871435e+01]
 [4.42998921e+01 7.02623043e+04 3.87072844e+04 2.01618123e+00
  3.38356197e+04 4.48208630e+01]]
Las desviaciones son:
[[1.47105594e+01 6.20696435e+03 3.21901567e+03 2.53514671e+00
  2.45685863e+03 1.41731672e+01]
 [1.50523273e+01 1.13412265e+05 6.59025571e+04 2.55740732e+00
  6.29913911e+04 1.54786058e+01]]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


**Análisis de Resultados**

De los resultados arrojados por el modelo, se observa que los clientes riesgosos tienen ingresos mayores a 70 mil USD y los clientes normales tienen ingresos de 6 mil USD aproximadamente. Esto quiere decir que los clientes riesgosos tienen ingresos mucho más grandes.

Pasa lo mismo con los gastos, donde los clientes riesgosos tienen gastos mucho mas altos que los clientes normales. Esto muestra un flujo de grandes cantidades de dinero por parte de los clientes riesgosos que indican una posible existencia de actividades inusuales.

Existe tambien una gran diferencia en la variable Monto transado con tarjetas, ya que los clientes riesgosos transan más de 33 mil USD y lo clientes normales solo alrededor de 3 mil USD. Esta gran diferencia se puede considerar como el indicador principal para reconocer lavado de activos.

Las otras variables (edad, número de tajetas y el porcentaje de crecimiento) tienen valores similares entre los clientes riesgosos y los normales, por lo cual son menos llamativas para el análisis de lavado de activos.

3. Se procede con la evaluación del modelo. Para la evaluación de este tipo de modelo se utiliza la matriz de confusión.

In [ ]:
# Predecir cosas a traves de la matriz XD
ydp=mnb.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)

# Se determinan las metricas de la matriz de confusión
VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]; TDatos=len(XDB)

# 1. Exactitud: Funcionamiento general del modelo
Ex=(VP+VN)/TDatos
print('Exactitud: ',Ex)

# 2. Tasa de Error: Es el porcentaje de fallo del sistema (fallos del modelo)
TEr=(FP+FN)/TDatos
print('Tasa de Error: ',TEr)

# 3. Sensibilidad: Como se comportó con respecto a los positivos SOLAMENTE
Sen=VP/(VP+FN)
print('Sensibilidad: ',Sen)

# 4. Especificidad: Como se comporta el modelo pronosticando negativos
Esp=VN/(VN+FP)
print('Especificidad: ',Esp)

# 5. Precisión: Es una versión de como se comporta el modelo, frente a los positivos SOLAMENTE
Pre=VP/(VP+FP)
print('Precisión: ',Pre)

# 6. Predicción Negativa: Como funciona el modelo pronosticando créditos PreNegados
PrN=VN/(VN+FN)
print('Predicción Negativa: ',PrN)


[[2179   44]
 [ 145  782]]
Exactitud:  0.94
Tasa de Error:  0.06
Sensibilidad:  0.8435814455231931
Especificidad:  0.9802069275753487
Precisión:  0.9467312348668281
Predicción Negativa:  0.9376075731497419


**Análisis de resultados**

De acuerdo con los resultados, el modelo tiene una EXACTITUD del 94%, esto indica el buen comportamiento del modelo en cuestión de clasificación de los clientes (normales y riesgosos).

La TASA DE ERROR fue del 6%, lo que quiere decir que las clasificaciones incorrectas fueron menos, esto quiere decir que el modelo comete pocos errores al identificar clientes normales y riesgosos.

Por otro lado, la SENSIBILIDAD fue alrededor del 84%, lo que quiere decir que el modelo tiene la cpacidad de identificar de manera correcta la mayor parte de casos riesgosos. Sin embargo, aun existen 145 casos que no se lograron identificar, pero al mismo tiempo, al ser tan pocos casos es fácil corregir este tipo de errores.


La ESPECIFICIDAD resultó ser del 98% lo que quiere decir que la clasificación de los clientes normales, lo quue signfica un buen desemepño operativo.

Un 94% de PRECISIÓN quiere decir que si el modelo identifica un cliente riesgo, significa que es verdadero.

Por último, la PREDICCIÓN NEGATIVA dejó como resultado casi un 94%, que significa que el modelo clasifica a los clientes normales de manera correcta y verifica que no representen un riesgo

4. Evaluamos un solicitante

In [ ]:
XP=[41,7000,4000,2,3000,38]
ydc=mnb.predict([XP])
print(ydc)

if ydc==1:
  print("El cliente es riesgoso por lavado de activos ")
else:
  print("El cliente es normal")

[0]
El cliente es normal


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(
